pkg/api/dirs_test.go

```diff
@@ -0,0 +1,247 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package api_test
+
+import (
+	"archive/tar"
+	"bytes"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path"
+	"testing"
+
+	"github.com/ethersphere/bee/pkg/api"
+	"github.com/ethersphere/bee/pkg/jsonhttp"
+	"github.com/ethersphere/bee/pkg/jsonhttp/jsonhttptest"
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/manifest/jsonmanifest"
+	"github.com/ethersphere/bee/pkg/storage/mock"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bee/pkg/tags"
+)
+
+type dirTestCase struct {
+	expectedHash string
+	files        []dirTestCaseFile
+}
+
+type dirTestCaseFile struct {
+	data      []byte
+	name      string
+	path      string
+	reference swarm.Address
+	headers   http.Header
+}
+
+func TestDirs(t *testing.T) {
+	var (
+		dirUploadResource    = "/dirs"
+		fileDownloadResource = func(addr string) string { return "/files/" + addr }
+		client               = newTestServer(t, testServerOptions{
+			Storer: mock.NewStorer(),
+			Tags:   tags.NewTags(),
+			Logger: logging.New(ioutil.Discard, 5),
+		})
+	)
+
+	t.Run("empty request body", func(t *testing.T) {
+		jsonhttptest.ResponseDirectSendHeadersAndReceiveHeaders(t, client, http.MethodPost, dirUploadResource, bytes.NewReader(nil), http.StatusBadRequest, jsonhttp.StatusResponse{
+			Message: "could not extract dir info from request",
+			Code:    http.StatusBadRequest,
+		}, nil)
+	})
+
+	t.Run("empty file", func(t *testing.T) {
+		f := writeAndOpenFile(t, "empty-file", nil)
```

Comment:
it is not entirely transparent where these files are created. it would be good to explicitly write to a [temp directory](https://golangcode.com/creating-temp-files/)

Reply:
this was my initial approach; the problem with this is that temp dirs and files use names with randomly generated appendices. this is good for concurrency.

the reason that this is a problem is that we cannot predict the hash of these temp assets. 

so the alternatives i see right now are:

- leave as is and have the test verify the resulting hashes
- change to explicitly temp files, and do not verify hashes
- somehow verify the hashes for each generated name (no idea how this could be done or even _if_ it could be done)

i don't know that there is a way to generate _explicitely_ temp files and dirs without a randomly generated name other than what i've done here. although if we keep it this way, we lose the benefit of better concurrency handling in case of parallel testing. not sure how much weight to put into this though.

some direction on which alternative to take here would be much appreciated :pray:

Reply:
this is now somewhat of a moot point due to [the most recent changes](https://github.com/ethersphere/bee/pull/447#discussion_r461297210)